<a href="https://colab.research.google.com/github/er-prateek-tripathi/Machine_Learning/blob/master/IBM%20AI%20Engineering/Course-2-Introduction%20to%20Deep%20Learning%20and%20Neural%20Networks%20with%20Keras%20/Regression%20With%20Keras/Regression_model_with_comprehensive_testing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1 align=center><font size = 5>Regression Models with Keras</font></h1>


## Introduction


Despite the popularity of more powerful libraries such as PyTorch and TensorFlow, they are not easy to use and have a steep learning curve. So, for people who are just starting to learn deep learning, there is no better library to use other than the Keras library.

Keras is a high-level API for building deep learning models. It has gained favor for its ease of use and syntactic simplicity facilitating fast development. As you will see in this lab and the other labs in this course, building a very complex deep learning network can be achieved with Keras with only few lines of code. You will appreciate Keras even more, once you learn how to build deep models using PyTorch and TensorFlow in the other courses.

So, in this lab, you will learn how to use the Keras library to build a regression model.


<h2>Regression Models with Keras</h2>

<h3>Objective for this Notebook<h3>    
<h5> 1. How to use the Keras library to build a regression model.</h5>
<h5> 2. Download and Clean dataset </h5>
<h5> 3. Build a Neural Network </h5>
<h5> 4. Train and Test the Network. </h5>     



## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>
    
1. <a href="#item31">Download and Clean Dataset</a>  
2. <a href="#item32">Import Keras</a>  
3. <a href="#item33">Build a Neural Network</a>  
4. <a href="#item34">Train and Test the Network</a>  

</font>
</div>


<a id="item31"></a>


## Download and Clean Dataset


Let's start by importing the <em>pandas</em> and the Numpy libraries.


In [17]:

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import Adam
from sklearn.metrics import mean_squared_error


<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>


Let's download the data and read it into a <em>pandas</em> dataframe.


In [2]:
!curl https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv -o concrete_data.csv

concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 58988  100 58988    0     0   280k      0 --:--:-- --:--:-- --:--:--  281k


Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa.


#### Let's check how many data points we have.


In [4]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.


Let's check the dataset for any missing values.


In [3]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.


# Part A
Build a baseline model

Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the
train_test_split
helper function from Scikit-learn.

2. Train the model on the training data using 50 epochs.

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.


In [19]:
# Load the concrete dataset
data = pd.read_csv('concrete_data.csv')

In [20]:
# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(data.drop('Strength', axis=1), data['Strength'], test_size=0.3)


In [21]:
# Create the neural network model
model = Sequential()
model.add(Dense(10, activation='relu', input_shape=(8,)))
model.add(Dense(1))


In [22]:
# Compile the model
model.compile(optimizer=Adam(), loss='mean_squared_error')

In [ ]:
# Train the model
model.fit(X_train, y_train, epochs=50)

In [26]:
# Evaluate the model
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)


10/10 [==============================] - 0s 2ms/step


In [27]:
# Print the mean and standard deviation of the mean squared errors
print('Mean:', np.mean(mse))
print('Standard deviation:', np.std(mse))

Mean: 113.60228258891229
Standard deviation: 0.0


In [30]:
mse_list = []

In [ ]:
# Repeat steps 1-3, 50 times
for i in range(50):
  # Split the data into training and test sets
  X_train, X_test, y_train, y_test = train_test_split(data.drop('Strength', axis=1), data['Strength'], test_size=0.3)

  # Create the neural network model
  model = Sequential()
  model.add(Dense(10, activation='relu', input_shape=(8,)))
  model.add(Dense(1))

  # Compile the model
  model.compile(optimizer=Adam(), loss='mean_squared_error')

  # Train the model
  model.fit(X_train, y_train, epochs=50)

  # Evaluate the model
  y_pred = model.predict(X_test)
  mse = mean_squared_error(y_test, y_pred)

  # Add the mean squared error to a list
  mse_list.append(mse)

# Print the mean and standard deviation of the mean squared errors
print('Mean:', np.mean(mse_list))
print('Standard deviation:', np.std(mse_list))

In [32]:
print('Mean:', np.mean(mse_list))
print('Standard deviation:', np.std(mse_list))

Mean: 475.19700013312877
Standard deviation: 511.2835052497344


# Part B

Normalize the data

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

How does the mean of the mean squared errors compare to that from Step A?

In [ ]:


# Load the concrete dataset
data = pd.read_csv('concrete_data.csv')

# Normalize the data
#data = data.drop('Strength', axis=1)
data = data.apply(lambda x: (x - x.mean()) / x.std())

# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(data, data['Strength'], test_size=0.3)

# Create the neural network model
model = Sequential()
model.add(Dense(10, activation='relu', input_shape=(9,)))
model.add(Dense(1))

# Compile the model
model.compile(optimizer=Adam(), loss='mean_squared_error')

# Train the model
model.fit(X_train, y_train, epochs=50)

# Evaluate the model
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)

# Print the mean and standard deviation of the mean squared errors


In [39]:
print('Mean:', np.mean(mse))
print('Standard deviation:', np.std(mse))

Mean: 0.019736591880074858
Standard deviation: 0.0


# Part C

Increate the number of epochs

Repeat Part B but use 100 epochs this time for training.

How does the mean of the mean squared errors compare to that from Step B?

In [ ]:


# Load the concrete dataset
data = pd.read_csv('concrete_data.csv')

# Normalize the data
#data = data.drop('Strength', axis=1)
data = data.apply(lambda x: (x - x.mean()) / x.std())

# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(data, data['Strength'], test_size=0.3)

# Create the neural network model
model = Sequential()
model.add(Dense(10, activation='relu', input_shape=(9,)))
model.add(Dense(1))

# Compile the model
model.compile(optimizer=Adam(), loss='mean_squared_error')

# Train the model
model.fit(X_train, y_train, epochs=100)

# Evaluate the model
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)


In [41]:
print('Mean:', np.mean(mse))
print('Standard deviation:', np.std(mse))

Mean: 0.005093594840284103
Standard deviation: 0.0


# Part D

Increase the number of hidden layers

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

How does the mean of the mean squared errors compare to that from Step B?


In [ ]:


# Load the concrete dataset
data = pd.read_csv('concrete_data.csv')

# Normalize the data
#data = data.drop('Strength', axis=1)
data = data.apply(lambda x: (x - x.mean()) / x.std())

# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(data, data['Strength'], test_size=0.3)

# Create the neural network model
model = Sequential()
model.add(Dense(10, activation='relu', input_shape=(9,)))
model.add(Dense(10, activation='relu'))
model.add(Dense(10, activation='relu'))
model.add(Dense(10, activation='relu'))
model.add(Dense(1))

# Compile the model
model.compile(optimizer=Adam(), loss='mean_squared_error')

# Train the model
model.fit(X_train, y_train, epochs=100)

# Evaluate the model
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)


In [43]:
print('Mean:', np.mean(mse))
print('Standard deviation:', np.std(mse))

Mean: 0.004491933476648155
Standard deviation: 0.0
